sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1300 @@
+events, then the behaviour is well-defined, and will match the observable
+total order.
+
+The returned value from the `info::queue::state` should be considered
+immediately stale in multi-threaded usage, as another thread could have
+preemptively changed the state of the queue.
+
+=== Exception Safety
+
+In addition to the destruction semantics provided by the SYCL
+{crs}[common reference semantics], when a modifiable `command_graph` is
+destroyed recording is ended on any queues that are recording to that
+graph, equivalent to `this->end_recording()`.
+
+As a result users don't need to manually wrap queue recording code in a
+`try` / `catch` block to reset the state of recording queues on an exception
+back to the executing state. Instead, an uncaught exception destroying the
+modifiable graph will perform this action, useful in RAII pattern usage.
+
+=== Error Handling
+
+Errors are reported through exceptions, as usual in the SYCL API. For new APIs,
+submitting a graph for execution can generate unspecified asynchronous errors,
+while `command_graph::finalize()` may throw unspecified synchronous exceptions.
+Synchronous exception errors codes are defined for all of
+`command_graph::begin_recording()`, `command_graph::end_recording()` and
+`command_graph::update()`.
+
+When a queue is in recording mode asynchronous exceptions will not be
+generated, as no device execution is occurring. Synchronous errors specified as
+being thrown in the default queue executing state, will still be thrown when a
+queue is in the recording state.
+
+The `command_graph::begin_recording` and `command_graph::end_recording`
+entry-points return a `bool` value informing the user whether a related queue
+state change occurred. False is returned rather than throwing an exception when
+no queue state is changed. This design is because the queues are already in
+the state the user desires, so if the function threw an exception in this case,
+the application would likely swallow it and then proceed.
+
+While a queue is in the recording state, methods performed on that queue which
+are not command submissions behave as normal. This includes waits, throws, and
+queries on the queue. These are all ignored by the graph system, as opposed to
+throwing an exception when in queue recording mode. This is because otherwise
+there would be no thread safe way for a user to check they could call these
+functions without throwing, as a query about the state of the queue may be
+immediately stale.
+
+=== Storage Lifetimes
+
+The lifetime of any buffer recorded as part of a submission
+to a command graph will be extended in keeping with the common reference
+semantics and buffer synchronization rules in the SYCL specification. It will be
+extended either for the lifetime of the graph (including both modifiable graphs
+and the executable graphs created from them) or until the buffer is no longer
+required by the graph (such as after being replaced through executable graph update).
```

Comment:
@gmlueck Starting a discussion about buffer lifetimes here since the other thread was quite long already. After our previous discussion I've been looking into the issue and I think really there are two concerns:

- Providing an error to users when they or library code try to use the buffer destructor copy-back mechanism.
- Extending storage lifetime to support the use of temporary buffers inside a graph.

For the first we discussed throwing an error when the copy-back mechanism is triggered but I think this may be problematic. The SYCL spec does state that, "If any error occurs on buffer destruction, it is reported via the associated queue’s asynchronous error handling mechanism.". However, I think the wording here is not sufficient, not just in this case but also for any errors thrown during copy-back to host. It makes sense for buffers with no attached host storage as the spec states their lifetimes are entirely managed by the runtime, where they are likely to be deleted by a queue once it has finished with the backend resources. In that case it can report the errors through its async_handler, but if the error happens in user code there is no queue/async_handler available to report these errors. In practice it seems in DPC++ any errors are just caught and only reported through the tracing API if enabled anyway.

We could also just throw the error without catching it in the destructor but I'm wary of that because throwing from a destructor can easily cause a `terminate` if it happens during stack unwinding among other bad behaviours. I see a couple of potential solutions though:

- We could require that buffers used inside a graph must have copy-back disabled (via `buffer::set_final_data()` or `buffer::set_copy_back()`) that way we should be able to throw an error at submission or finalize time if an incompatible buffer is used. This would give users an easily actionable error for their own code, and a meaningful error for incompatible library calls. Although it may restrict some library calls that use buffers that would trigger a copy-back but don't actually rely on this mechanism to do anything bad that couldn't be captured in a graph.
  - We could also just disallow buffers with any attached host storage but that may be too restrictive since we only need to disable copy-back. It may however be more clear and explicit to users, but I'm not so sure about that.
- We could continue documenting this behaviour in the specification, but make it more explicit about this behaviour and the implications of it being used inside library calls etc. This would give us the most potential library compatibility and give users enough information to be aware of the pitfalls in their own code and library code. However, we also can't guarantee that people will actually read the documentation before using it so it could lead to confusing behaviour.

I do think that extending buffer lifetime is a separate issue, but I feel it is compatible with both solutions and is important for capturing code where temporary buffers are used:
- If we require disabling copy-back there should be no issue with extending lifetimes because the user is not expecting a copy-back to happen anyway.
- If we commit to just documenting the behaviour and implications in detail then it should be easy to reason about when reading the documentation.

Tagging @reble for any thoughts also.

Reply:
> We could also just disallow buffers with any attached host storage but that may be too restrictive since we only need to disable copy-back. It may however be more clear and explicit to users, but I'm not so sure about that.

Some variant on this might make sense.  However, the rule would be "disallow any command that has an accessor on a buffer which would cause a write-back to happen".  This would disallow some buffers with attached memory, but not all.  In particular, the following cases would not be disallowed:

* A command that has no writeable accessors on the buffer.
* A buffer that takes a `const` pointer to attached memory.
* A buffer that takes a `shared_ptr` to host memory, when the application has released ownership of that memory by the time the command was submitted.

I'm not sure if this would be too restrictive.  Do you have some typical uses cases in mind?  Would those use cases be prevented from these rules?

I though of another case that needs to be flagged too.  Here is the complete list I have thought of so far:

* Adding a command with an accessor on a buffer that would trigger a write-back (defined above).
* Waiting on the queue or on an event from a recorded command.
* Constructing a host accessor to a buffer that has a writeable accessor used in a recorded command.

(I think I missed the host accessor case before.)

> In practice it seems in DPC++ any errors are just caught and only reported through the tracing API if enabled anyway.

Are you saying that DPC++ ignores all backend errors unless the tracing API is enabled?  That would be really bad ...

Reply:
> Some variant on this might make sense. However, the rule would be "disallow any command that has an accessor on a buffer which would cause a write-back to happen". This would disallow some buffers with attached memory, but not all. 

Yes I think that makes sense, and does allow the most compatibility as you say.

> I'm not sure if this would be too restrictive. Do you have some typical uses cases in mind? Would those use cases be prevented from these rules?

I think it is only as restrictive as it needs to be. The only case I can see that would be prevented is probably really a programming error regardless of the rule: where a copy-back is triggered but the application or library isn't actually doing anything with the copied data, so it wouldn't cause us any problems. I don't think there's any way to support that though, nor probably should we.

> Constructing a host accessor to a buffer that has a writeable accessor used in a recorded command.

I agree this is a problematic case, though it does present some implementation challenges since the `host_accessor` doesn't interact with the graph or a queue. I would guess it necessitates some awareness in the buffer of it being part of a graph command (which would also have been an issue with detecting this in the buffer destructor).

> Are you saying that DPC++ ignores all backend errors unless the tracing API is enabled? That would be really bad ...

Yes, but apologies I actually misread it slightly. The errors are in fact just swallowed in the destructor (with the commit citing the previously discussed dangers of throwing inside destructors) and not reported through the tracing API. The tracing API just indicates that the buffer destructor itself was called. [Seen here](https://github.com/intel/llvm/blob/f350461f7e807fafd83acf8784da87cea2677cbc/sycl/source/detail/buffer_impl.hpp#L158)

Reply:
> The errors are in fact just swallowed in the destructor (with the commit citing the previously discussed dangers of throwing inside destructors) 

Thanks for bringing this to my attention.  I reported this internally.

Reply:
> Constructing a host accessor to a buffer that has a writeable accessor used in a recorded command.

I had some more thoughts about this case, and I'm not sure if it makes sense to block them altogether but perhaps only during recording (though this does increase the burden of tracking this state in the buffer). I'm considering a scenario like this:

```c++
queue.submit([&](handler& cgh){
  cgh.ext_oneapi_graph(execGraph);
});

// User constructs a host_accessor to access results from the graph in between executions
host_accessor hostAcc(someBufferUsedInTheGraph);

// User re-submits the graph
queue.submit([&](handler& cgh){
  cgh.ext_oneapi_graph(execGraph);
});
```

This seems to me like a case that should be allowed, since the `host_accessor` triggers an explicit copy-back in contrast to the destructor mechanism. So perhaps the spec should be worded only to disallow a host-accessor on buffers which would trigger a write-back and which are being currently recorded to a graph as part of some command.

Reply:
Yes, I agree that should work.  All of the cases I list above should only throw an exception when the queue is in record mode.  For example, it's an error to wait on a queue in record mode, but it's OK to wait on a queue in non-record mode.

As you note, the implementation will need to do some bookkeeping to know when there is an accessor to a buffer where the accessor is added to a command in a record-mode queue.  Once the queue goes out of record-mode, it will once again be OK to create a host accessor on the buffer.